upstairs/src/volume.rs

```diff
@@ -1092,15 +1094,355 @@ impl Volume {
             }
         }
     }
+
+    // We compare two VolumeConstructionRequests to see if the second one
+    // can be a replacement for the first.
+    // The requirements are:
+    // 1. Only VolumeConstructionRequests::Volume type is supported.
+    // 2. Sub volumes must all be VolumeConstructionRequest::Region
+    // 3. Everything must be the same between the two Volumes, except:
+    //    A. The new generation number must be greater than the old.
+    //    B. Only one CrucibleOpts::target[] is different.
+    //    C. The new Volume can have None for read only parent if the
+    //       original had Some(), or it must match the Some().
+    //
+    // Any other difference between the two volumes is considered a failure.
+    //
+    // On success, we record the old and new downstairs target and return
+    // those to the caller.
+    pub async fn compare_vcr_for_replacement(
```

Comment:
A high-level thought: would it make sense to use this during live migration to ensure that two Crucible backends have been configured compatibly? (You'd have the old VCR on the source and the new VCR on the target; the source would send its VCR to the target; the target could then verify that its VCR is plausibly compatible with the one on the source.) We'd have a little work to do in Propolis to enable this but it's nothing complicated.

The reason I ask is that if this *could* be used this way, it may be nicer to phrase the routine name and error messages throughout in terms of "compatibility" instead of "replaceability."

Reply:
I think this could be a more generic check, with a little more polish.

For a migration, the generation on the new side should be > than the old, which we check here.  For a migration, the read only parent could be `None` on the new side, which we check here.  I'm assuming we don't want to migrate and replace a downstairs at the same time (at least initailly).

To make this generic, it would have to differentiate between the "okay for migration" VCR and "ok for downstairs replacement" VCR.

I think a way to do that would be to return `Ok(None)` if the VCRs are the same (given the above constraints) and 'Ok(Some(old_target, new_target))` if the VCRs are the same with the exception of one target being different.

How does that sound?